Introduce setting `create_table_empty_primary_key_by_default`

srikanthccv opened the pull request:
Fixes #55023 
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Introduce setting `create_table_empty_primary_key_by_default` for default `ORDER BY ()`

### Documentation entry for user-facing changes

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->
If set to true and the table of MergeTree family does not contain ORDER BY and PRIMARY KEY, then an empty tuple is used by default


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

srikanthccv committed:
Introduce setting `create_table_empty_primary_key_by_default`

If set to true and the table of MergeTree family does not contain
ORDER BY and PRIMARY KEY, then an empty tuple is used by default

srikanthccv committed:
Update merge tree doc

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #55899 from srikanthccv/ClickHouse:issue-55023 -->
*This is an automated comment for commit 0f2e0ade996a9c723b13676d44d168e9a955b3fd with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55899/0f2e0ade996a9c723b13676d44d168e9a955b3fd/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6598429823">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55899/0f2e0ade996a9c723b13676d44d168e9a955b3fd/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

alexey-milovidov commented:
Need to add tests.

alexey-milovidov commented:
Tests should include:
- a check that it throws an exception when there is no primary key, and the setting is not enabled;
- a check that it successfully creates a table when there is no primary key, and the setting is enabled;
- a check that it does not get confused when you specify a primary on a per-column basis (x String PRIMARY KEY, y UInt64 PRIMARY KEY) and the setting is enabled;

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into issue-55023

srikanthccv committed:
Add integration tests for create_table_empty_primary_key_by_default

alexey-milovidov commented:
Functional tests should be sufficient; no need for integration tests. See `tests/queries/0_stateless`.

srikanthccv commented:
I was unsure about that but yes, makes sense. Let me update.

srikanthccv committed:
Revert "Add integration tests for create_table_empty_primary_key_by_default"

This reverts commit 754cfdc68c4582c95d3f1274b9d5223ca599da96.

srikanthccv committed:
Add functional tests for create_table_empty_primary_key_by_default

1. 02902_empty_order_by_throws_error
   - setting disabled and no order by or primary key - expect error
   - setting disabled and primary key in the table definition - no error and expect order by as primary key
2. 02903_empty_order_by_with_setting_enabled
   - setting enabled and no order by or primary key - expect the order by as empty
   - setting enabled and per-column primary key - expect the order by as column order
   - setting enabled and primary key in table definition (not per-column or order by) - expect order as primary key
   - setting enabled and order by in table definition (no primary key) - expect same order by

Avogar commented on docs/en/engines/table-engines/mergetree-family/mergetree.md:
```diff
@@ -73,7 +73,7 @@ A tuple of column names or arbitrary expressions. Example: `ORDER BY (CounterID,
 ClickHouse uses the sorting key as a primary key if the primary key is not defined explicitly by the `PRIMARY KEY` clause.
 
-Use the `ORDER BY tuple()` syntax, if you do not need sorting. See [Selecting the Primary Key](#selecting-the-primary-key).
+Use the `ORDER BY tuple()` syntax, if you do not need sorting Or set `create_table_empty_primary_key_by_default` to `true` to use the `ORDER BY tuple()` syntax by default. See [Selecting the Primary Key](#selecting-the-primary-key).
```
```suggestion
Use the `ORDER BY tuple()` syntax, if you do not need sorting, or set `create_table_empty_primary_key_by_default` to `true` to use the `ORDER BY tuple()` syntax by default. See [Selecting the Primary Key](#selecting-the-primary-key).
```

Avogar committed:
Update docs/en/engines/table-engines/mergetree-family/mergetree.md

Avogar commented:
Seems like some problem with CI, please, update your branch with master. We cannot merge PR with failed build checks

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into issue-55023

srikanthccv committed:
Merge branch 'issue-55023' of github.com:srikanthccv/ClickHouse into issue-55023

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into issue-55023

srikanthccv committed:
Rename the file number since master already has it
